File: vmidentity/openidconnect/protocol/src/main/java/com/vmware/identity/openidconnect/protocol/JSONUtils.java
/*
 *  Copyright (c) 2012-2015 VMware, Inc.  All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
 *  use this file except in compliance with the License.  You may obtain a copy
 *  of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, without
 *  warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 *  License for the specific language governing permissions and limitations
 *  under the License.
 */

package com.vmware.identity.openidconnect.protocol;

import java.net.URI;

import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;

import org.apache.commons.lang3.Validate;

import com.nimbusds.jose.util.JSONObjectUtils;
import com.nimbusds.jwt.SignedJWT;
import com.vmware.identity.openidconnect.common.ParseException;

/**
 * @author Yehia Zayour
 */
public final class JSONUtils {
    private JSONUtils() {
    }

    public static long getLong(JSONObject json, String key) throws ParseException {
        Validate.notNull(json, "json");
        Validate.notEmpty(key, "key");

        Object objectValue = json.get(key);
        if (objectValue == null) {
            throw new ParseException(String.format("json is missing %s member", key));
        }

        if (!(objectValue instanceof Long)) {
            throw new ParseException(String.format("json has non-long %s member", key));
        }

        Long longValue = (Long) objectValue;
        return longValue.longValue();
    }

    public static String getString(JSONObject json, String key) throws ParseException {
        Validate.notNull(json, "json");
        Validate.notEmpty(key, "key");

        Object objectValue = json.get(key);
        if (objectValue == null) {
            throw new ParseException(String.format("json is missing %s member", key));
        }

        if (!(objectValue instanceof String)) {
            throw new ParseException(String.format("json has non-string %s member", key));
        }

        String stringValue = (String) objectValue;
        if (stringValue.isEmpty()) {
            throw new ParseException(String.format("json has empty %s member", key));
        }

        return stringValue;
    }

    public static String[] getStringArray(JSONObject json, String key) throws ParseException {
        Validate.notNull(json, "json");
        Validate.notEmpty(key, "key");

        Object objectValue = json.get(key);
        if (objectValue == null) {
            throw new ParseException(String.format("json is missing %s member", key));
        }

        if (!(objectValue instanceof JSONArray)) {
            throw new ParseException(String.format("json has non-array %s member", key));
        }

        String[] stringArrayValue;
        try {
            stringArrayValue = ((JSONArray) objectValue).toArray(new String[0]);
        } catch (ArrayStoreException e) {
            throw new ParseException(String.format("json has non-string-array %s member", key), e);
        }

        if (stringArrayValue.length == 0) {
            throw new ParseException(String.format("json has empty %s member", key));
        }

        return stringArrayValue;
    }

    public static URI getURI(JSONObject json, String key) throws ParseException {
        Validate.notNull(json, "json");
        Validate.notEmpty(key, "key");

        String stringValue = JSONUtils.getString(json, key);
        try {
            return URIUtils.parseURI(stringValue);
        } catch (ParseException e) {
            throw new ParseException(String.format("json has invalid %s member", key), e);
        }
    }

    public static SignedJWT getSignedJWT(JSONObject json, String key) throws ParseException {
        Validate.notNull(json, "json");
        Validate.notEmpty(key, "key");

        String stringValue = JSONUtils.getString(json, key);
        try {
            return SignedJWT.parse(stringValue);
        } catch (java.text.ParseException e) {
            throw new ParseException(String.format("failed to parse %s member", key), e);
        }
    }

    public static JSONObject parseJSONObject(String jsonString) throws ParseException {
        Validate.notEmpty(jsonString, "jsonString");

        try {
            return JSONObjectUtils.parseJSONObject(jsonString);
        } catch (java.text.ParseException e) {
            throw new ParseException("failed to parse JSONObject", e);
        }
    }
}

File: mutentca/server/common/jsonutils.c
/*
 * Copyright © 2018 VMware, Inc.  All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the “License”); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS, without
 * warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

#include <includes.h>

DWORD
LwCAJsonLoadObjectFromFile(
    PCSTR                   pcszFilePath,
    PLWCA_JSON_OBJECT       *ppJsonConfig
    )
{
    DWORD                   dwError = 0;
    json_error_t            jsonError = {0};
    PLWCA_JSON_OBJECT       pJsonConfig = NULL;
    size_t                  szJsonFlags = JSON_DECODE_ANY;

    if (IsNullOrEmptyString(pcszFilePath) || !ppJsonConfig)
    {
        dwError = LWCA_ERROR_INVALID_PARAMETER;
        BAIL_ON_LWCA_ERROR(dwError);
    }

    if (!(pJsonConfig = json_load_file(pcszFilePath, szJsonFlags, &jsonError)))
    {
        LWCA_LOG_ERROR(
                "[%s,%d] Failed to open config file (%s). Error: %s:%d",
                __FUNCTION__,
                __LINE__,
                pcszFilePath,
                jsonError.text,
                jsonError.line);
        dwError = LWCA_JSON_FILE_LOAD_ERROR;
        BAIL_ON_LWCA_ERROR(dwError);
    }

    *ppJsonConfig = pJsonConfig;


cleanup:

    return dwError;

error:

    LwCAJsonCleanupObject(pJsonConfig);
    if (ppJsonConfig)
    {
        *ppJsonConfig = NULL;
    }

    goto cleanup;
}

DWORD
LwCAJsonGetObjectFromKey(
    PLWCA_JSON_OBJECT       pJson,
    PCSTR                   pcszKey,
    PLWCA_JSON_OBJECT       *ppJsonValue
    )
{
    DWORD                   dwError = 0;
    PLWCA_JSON_OBJECT       pJsonValue = NULL;

    if (!pJson || IsNullOrEmptyString(pcszKey) || !ppJsonValue)
    {
        dwError = LWCA_ERROR_INVALID_PARAMETER;
        BAIL_ON_LWCA_ERROR(dwError);
    }

    if (!(pJsonValue = json_object_get(pJson, pcszKey)))
    {
        dwError = LWCA_JSON_PARSE_ERROR;
        BAIL_ON_JSON_PARSE_ERROR(dwError);
    }

    *ppJsonValue = pJsonValue;

cleanup:

    return dwError;

error:

    if (ppJsonValue)
    {
        *ppJsonValue = NULL;
    }

    goto cleanup;
}

DWORD
LwCAJsonGetStringFromKey(
    PLWCA_JSON_OBJECT       pJson,
    PCSTR                   pcszKey,
    PSTR                    *ppszValue
    )
{
    DWORD                   dwError = 0;
    PLWCA_JSON_OBJECT       pJsonValue = NULL;
    PCSTR                   pcszValue = NULL;

    if (!pJson || IsNullOrEmptyString(pcszKey) || !ppszValue)
    {
        dwError = LWCA_ERROR_INVALID_PARAMETER;
        BAIL_ON_LWCA_ERROR(dwError);
    }

    dwError = LwCAJsonGetObjectFromKey(pJson, pcszKey, &pJsonValue);
    BAIL_ON_LWCA_ERROR(dwError);

    pcszValue = json_string_value(pJsonValue);
    if (IsNullOrEmptyString(pcszValue))
    {
        dwError = LWCA_JSON_PARSE_ERROR;
        BAIL_ON_LWCA_ERROR(dwError);
    }

    dwError = LwCAAllocateStringA(pcszValue, ppszValue);
    BAIL_ON_LWCA_ERROR(dwError);

cleanup:

    return dwError;

error:

    if (ppszValue)
    {
        *ppszValue = NULL;
    }

    goto cleanup;
}

VOID
LwCAJsonCleanupObject(
    PLWCA_JSON_OBJECT       pJson
    )
{
    if (pJson)
    {
        json_decref(pJson);
    }
}
